lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -418,6 +425,62 @@ class LowerXMRPass : public LowerXMRBase<LowerXMRPass> {
     return success();
   }
 
+  LogicalResult handleTopModuleRefPorts() {
+    auto topModule = dyn_cast<FModuleOp>(*circuitOp.getMainModule());
+    if (!topModule)
+      return success();
+
+    auto builder = ImplicitLocOpBuilder::atBlockBegin(topModule.getLoc(),
+                                                      getOperation().getBody());
+
+    for (size_t portIndex = 0, numPorts = topModule.getNumPorts();
+         portIndex != numPorts; ++portIndex) {
+      auto refType = topModule.getPortType(portIndex).dyn_cast<RefType>();
+      if (!refType || isZeroWidth(refType.getType()) ||
+          topModule.getPortDirection(portIndex) != Direction::Out)
+        continue;
+      auto portValue = topModule.getArgument(portIndex);
+      auto remoteOpPath = getRemoteRefSend(portValue);
+      if (!remoteOpPath)
+        return failure();
+      SmallVector<Attribute> refSendPath;
+      SmallString<128> formatString;
+      size_t formatIndex = 0;
+      while (remoteOpPath) {
```

Comment:
This needs to be building up a hierpath rather than a series of piece-wise substitutions.